Coverage: ModuleNotFoundError: No module named 'html.parser'; 'html' is not a package

<!--🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅

Oh hi there! 😄 

To expedite issue processing please search open and closed issues before submitting a new one.
Existing issues often contain information about workarounds, resolution, or progress updates.

🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅-->


# 🐞 bug report

### Affected Rule
The issue is caused by the rule: 
`rules_python`

### Description
When activating coverage report in `rules_python` in a bazel workspace app that use Django, there is a clash in-between `coverage.py` and `django` that cause the `html` library to be re-imported

## 🔥 Exception or Error
```
==================================== ERRORS ====================================
________________________ ERROR collecting test session _________________________
../../../../../../../../../../../../execroot/__main__/external/python3_9_aarch64-apple-darwin/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1030: in _gcd_import
    ???
<frozen importlib._bootstrap>:1007: in _find_and_load
    ???
<frozen importlib._bootstrap>:986: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:680: in _load_unlocked
    ???
../pip_pytest/site-packages/_pytest/assertion/rewrite.py:170: in exec_module
    exec(co, module.__dict__)
account/tests/conftest.py:2: in <module>
    from cbporm.cbp.models import User
cbporm/__init__.py:21: in <module>
    django.setup()
../pip_django/site-packages/django/__init__.py:16: in setup
    from django.urls import set_script_prefix
../pip_django/site-packages/django/urls/__init__.py:1: in <module>
    from .base import (
../pip_django/site-packages/django/urls/base.py:8: in <module>
    from .exceptions import NoReverseMatch, Resolver404
../pip_django/site-packages/django/urls/exceptions.py:1: in <module>
    from django.http import Http404
../pip_django/site-packages/django/http/__init__.py:5: in <module>
    from django.http.response import (
../pip_django/site-packages/django/http/response.py:16: in <module>
    from django.core.serializers.json import DjangoJSONEncoder
../pip_django/site-packages/django/core/serializers/__init__.py:23: in <module>
    from django.core.serializers.base import SerializerDoesNotExist
../pip_django/site-packages/django/core/serializers/base.py:7: in <module>
    from django.db import models
../pip_django/site-packages/django/db/models/__init__.py:3: in <module>
    from django.db.models.aggregates import *  # NOQA
../pip_django/site-packages/django/db/models/aggregates.py:5: in <module>
    from django.db.models.expressions import Case, Func, Star, When
../pip_django/site-packages/django/db/models/expressions.py:10: in <module>
    from django.db.models import fields
../pip_django/site-packages/django/db/models/fields/__init__.py:11: in <module>
    from django import forms
../pip_django/site-packages/django/forms/__init__.py:6: in <module>
    from django.forms.boundfield import *  # NOQA
../pip_django/site-packages/django/forms/boundfield.py:4: in <module>
    from django.forms.utils import flatatt, pretty_name
../pip_django/site-packages/django/forms/utils.py:7: in <module>
    from django.utils.html import escape, format_html, format_html_join, html_safe
../pip_django/site-packages/django/utils/html.py:6: in <module>
    from html.parser import HTMLParser
E   ModuleNotFoundError: No module named 'html.parser'; 'html' is not a package

```

## 🌍  Your Environment

**Operating System:**

<pre>
  <code>
macOS (M1)
  </code>
</pre>

**Output of `bazel version`:**

<pre>
  <code>
Bazelisk version: development
Build label: 6.1.2
Build target: bazel-out/darwin_arm64-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Tue Apr 18 15:31:54 2023 (1681831914)
Build timestamp: 1681831914
Build timestamp as int: 1681831914
  </code>
</pre>

**Rules_python version:**

<pre>
  <code>
0.21.0
  </code>
</pre>

--- a/examples/bzlmod/test.py
+++ b/examples/bzlmod/test.py
@@ -12,12 +12,42 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import os
+import pathlib
+import sys
 import unittest
 
 from lib import main
 
 
 class ExampleTest(unittest.TestCase):
+    def test_coverage(self):
+        for i, path in enumerate(sys.path[1:-2]):
+            paths = ",\n    ".join(sys.path)
+
+            self.assertFalse(
+                "/coverage" in path,
+                f"Expected {i + 2}th '{path}' to not contain 'coverage.py' paths, "
+                f"sys.path has {len(sys.path)} items:\n    {paths}",
+            )
+
+        first_item, last_item = sys.path[0], sys.path[-1]
+        self.assertFalse(
+            first_item.endswith("coverage"),
+            f"Expected the first item in sys.path '{first_item}' to not be related to coverage",
+        )
+        if os.environ.get("COVERAGE_MANIFEST"):
+            # we are running under the 'bazel coverage :test'
+            self.assertTrue(
+                "pypi__coverage_cp" in last_item,
+                f"Expected {last_item} to be related to coverage",
+            )
+            self.assertEqual(pathlib.Path(last_item).name, "coverage")
+        else:
+            self.assertFalse(
+                "coverage" in last_item, f"Expected coverage tooling to not be present"
+            )
+
     def test_main(self):
         self.assertEquals(
             """\

--- a/examples/bzlmod/test.py
+++ b/examples/bzlmod/test.py
@@ -21,14 +21,29 @@ from lib import main
 
 
 class ExampleTest(unittest.TestCase):
-    def test_coverage(self):
-        for i, path in enumerate(sys.path[1:-2]):
-            paths = ",\n    ".join(sys.path)
+    def test_coverage_doesnt_shadow_stdlib(self):
+        # When we try to import the html module
+        import html as html_stdlib
 
+        try:
+            import coverage.html as html_coverage
+        except ImportError:
+            self.skipTest("not running under coverage, skipping")
+
+        self.assertNotEqual(
+            html_stdlib,
+            html_coverage,
+            "'html' import should not be shadowed by coverage",
+        )
+
+    def test_coverage_sys_path(self):
+        all_paths = ",\n    ".join(sys.path)
+
+        for i, path in enumerate(sys.path[1:-2]):
             self.assertFalse(
                 "/coverage" in path,
                 f"Expected {i + 2}th '{path}' to not contain 'coverage.py' paths, "
-                f"sys.path has {len(sys.path)} items:\n    {paths}",
+                f"sys.path has {len(sys.path)} items:\n    {all_paths}",
             )
 
         first_item, last_item = sys.path[0], sys.path[-1]

